### PR TITLE
Setting right TFM when running nuget.exe update for native or js projects

### PR DESCRIPTION
**Pull request by jainaashish**
Setting right TFM when running nuget.exe update for native or js projects. And using project extension to figure out project type like 'vcxproj' for c++ and 'jsproj' for javascript since msbuild projects don't have Project.Kind property.

Fix https://github.com/NuGet/Home/issues/1291

### CODE REVIEW

**Comment by jainaashish** [comment]
@yishaigalatzer @emgarten @alpaix

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 1]
```diff
@@ -46,17 +48,71 @@ public NuGetFramework TargetFramework
...
                 }
-                return NuGetFramework.Parse(moniker);
+
+                var framework = NuGetFramework.Parse(moniker);
+
+                // if the framework is .net core 4.5.1 return windows 8.1
+                if (framework.Framework.Equals(FrameworkConstants.FrameworkIdentifiers.NetCore)
+                    && framework.Version.Equals(Version.Parse("4.5.1.0")))
+                {
+                    return new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.Windows, Version.Parse("8.1"), framework.Profile);
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: you can just new a object Version instead of parsing

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 2]
```diff
@@ -46,17 +48,71 @@ public NuGetFramework TargetFramework
...
+
                 if (String.IsNullOrEmpty(moniker))
                 {
                     return null;
                 }
-                return NuGetFramework.Parse(moniker);
+
+                var framework = NuGetFramework.Parse(moniker);
+
+                // if the framework is .net core 4.5.1 return windows 8.1
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: can you explain in the comment why this change is necessary

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: also break long lines

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: and last question: Shouldn't this be a generic logic somewhere in frameworks? Why is it being repeated here?

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 3]
```diff
@@ -46,17 +48,71 @@ public NuGetFramework TargetFramework
...
+
+                return string.Format(CultureInfo.InvariantCulture, "{0}, Version={1}", platformIdentifier, platformVersion);
+            }
+
+            // Check for C++ project
+            if (StringComparer.OrdinalIgnoreCase.Equals(VCXProjextExt, extension))
+            {
+                // The C++ project does not have a TargetFrameworkMoniker property set. 
+                // We hard-code the return value to Native.
+                return "Native, Version=0.0";
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: isn't this already hardcoded elsewhere?

Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: It's as part of EnvDTEProjectUtility which is part of PackageManagement.VisualStudio project... I didn't want to create commandline dependency on this project neither i found a common place to keep this logic at one place.

More over figuring out project type is different at both the places so thought of keeping separate, but will love to get a suggestion to have a common code for this.

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs (hunk 1 above)
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: currently it does not exist any where in Frameworks, but i can add a new api something like GetCompatibleFramework(framwork) or ParseCompatibleFramework() in NugetFramework itself so that the same api could be used across vs, commandline,

Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs]: Makes sense, please work with Justin & Joel on the design

File: src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs [hunk 4]
```diff
@@ -37,6 +37,36 @@ var mostCompatibleGroup
...
         /// <summary>
+        /// Parse existing nuget framework for .net core 4.5.1 or 4.5 and return compatible framework instance
+        /// </summary>
+        /// <param name="framework"></param>
+        /// <returns></returns>
+        public static NuGetFramework GetCompatibleFramework(NuGetFramework framework)
+        {
+            if (framework == null)
+            {
+                throw new ArgumentNullException("framework");
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs:48]: nit: nameof(framework)

File: src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs [hunk 5]
```diff
@@ -37,6 +37,36 @@ var mostCompatibleGroup
         }
 
         /// <summary>
+        /// Parse existing nuget framework for .net core 4.5.1 or 4.5 and return compatible framework instance
+        /// </summary>
+        /// <param name="framework"></param>
+        /// <returns></returns>
+        public static NuGetFramework GetCompatibleFramework(NuGetFramework framework)
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs:44]: how about _GetProjectFrameworkReplacement_

**Comment by emgarten** [comment]
Address feedback then :shipit:

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 6]
```diff
@@ -46,17 +48,61 @@ public NuGetFramework TargetFramework
         {
             get
             {
-                string moniker = GetPropertyValue("TargetFrameworkMoniker");
+                // this is required to get the right TFM for native or js projects since TargetFrameworkMoniker
+                // property won't give the accurate value for these kind of projects
+                string moniker = GetTargetFrameworkString();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:53]: nit: `var`

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 7]
```diff
@@ -46,17 +48,61 @@ public NuGetFramework TargetFramework
...
+                // further parse framework for .net core 4.5.1 or 4.5 and get compatible framework instance
+                return MSBuildNuGetProjectSystemUtility.GetCompatibleFramework(framework);
             }
         }
 
         private dynamic Project { get; }
 
+        private string GetTargetFrameworkString()
+        {
+            var extension = GetPropertyValue("ProjectExt");
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:71]: all property names should be constants

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 8]
```diff
@@ -14,6 +14,8 @@ namespace NuGet.Common
     public class MSBuildProjectSystem : MSBuildUser, IMSBuildNuGetProjectSystem
     {
         private const string TargetName = "EnsureNuGetPackageBuildImports";
+        private const string JSProjectExt = ".jsproj";
+        private const string VCXProjextExt = ".vcxproj";
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:18]: would be `src/NuGet.Core/NuGet.ProjectManagement/Utility/Constants.cs` the right place for these constants?

File: src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs (hunk 4 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.ProjectManagement/Utility/MSBuildNuGetProjectSystemUtility.cs:48]: `nameof(framework)`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs [hunk 9]
```diff
@@ -1221,5 +1221,125 @@ public class NuGetUpdateCommandTests
                 Assert.True(content1.Contains(@"<HintPath>..\..\" + customPackageFolderName + @"\a.2.0.0\lib\net45\file.dll</HintPath>"));
             }
         }
+
+        [Fact]
+        public async Task UpdateCommand_Success_Native_JS_Projects()
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:1226]: As a recommendation, test name should follow the pattern [Given_When_Then](http://martinfowler.com/bliki/GivenWhenThen.html). 
In this case it would be named `UpdateCommand_Native_JS_Projects_Success`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs [hunk 10]
```diff
@@ -1221,5 +1221,125 @@ public class NuGetUpdateCommandTests
...
+                var r = CommandRunner.Run(
+                    Util.GetNuGetExePath(),
+                    workingPath,
+                    string.Join(" ", args),
+                    waitForExit: true);
+
+                Assert.True(r.Item1 == 0, "Output is " + r.Item2 + ". Error is " + r.Item3);
+
+                var content1 = File.ReadAllText(projectFile1);
+                Assert.True(content1.Contains(@"<Content Include=""test.txt"" />"));
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:1338]: `Assert.Contains`

**Comment by alpaix** [comment]
Minor style notes. Otherwise :shipit:

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs (hunk 8 above)
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:18]: make sense, i'll move those there

**Comment by jainaashish** [comment]
Merged 44d9169154dd0aa07df2cc24fe4536264c809106 after addressing comments.
